### ISSUE TITLE
Calling a view-like read-only function that does not alter the EVM state with ethers.js V6 does not work

### ISSUE THREAD

**Issue by iron4548** [hashgraph/hedera-json-rpc-relay#1858]
### Description

Uniswap V3 have read-only methods called `quoteExactInput` and `quoteExactOutput` to return a quote for a swap. They do not have the 'view' tag but they act like them. They always revert (but caught) and then return with a valid data.  They are intended to be used to get a quote and have no state change. So therefore I want to call these functions on the mirror node's EVM via a JSON-RPC relay.

See https://github.com/Uniswap/v3-periphery/blob/main/contracts/lens/QuoterV2.sol

I'm using ethers V6 to try and get a quote.

I tried to call the function without a signer, I get a error: "Error: contract runner does not support sending transactions (operation="sendTransaction", code=UNSUPPORTED_OPERATION, version=6.7.1)". This is most likely because ethers expects a signer due to the function not having a view tag.

(If I added the 'view' tag to the quoteExactInput function ABI - then it will work fine without a signer)

Anyhow, I added a signer and I got a 400 error (Error: server response 400 Bad Request) from the rpc relay. Since the error doesn't provide any useful tip or message ('undefined') to help me troubleshoot it myself, I created this issue to get some guidance. Why am I getting a 400 error?

```typescript

const networkId = 'testnet';
const provider = new ethers.JsonRpcProvider(`https://${networkId}.hashio.io/api`, '', {
  batchMaxCount: 1, //workaround for ethers V6
});

//note quoteExactInput does not have the 'view' tag, so a signer is required.
const interfaces = new ethers.Interface([
  'function quoteExactInput(bytes memory path, uint256 amountIn) external returns (uint256 amountOut, uint160[] memory sqrtPriceX96AfterList, uint32[] memory initializedTicksCrossedList, uint256 gasEstimate)'
]);

const privateKeyEd = process.env.TESTNET_OPERATOR_PRIVATEKEY_ED!;  
const signer = new ethers.Wallet(privateKeyEd, provider);

const quoterEvmAddress = `0x${ContractId.fromString(config.quoter).toSolidityAddress()}`;

const input =  100;
const inputAmountInSmallestUnit =  BigNumber(input * Math.pow(10, 8));
const token1 = TokenId.fromString(config.token1);
const token2 = TokenId.fromString(config..token2);

let pathHexStr = '0x';
pathHexStr += token1.toSolidityAddress();  
pathHexStr += feePercentToHexStr(0.03); //0.3% - swap fee
pathHexStr += token2.toSolidityAddress();

const encodedPathData = hexToUint8Array(pathHexStr);

const routerContract = new ethers.Contract(quoterEvmAddress, interfaces.fragments, signer);
const amounts = await routerContract.quoteExactInput(encodedPathData, inputAmountInSmallestUnit.toString()); //fails with a 400
```

Is there a proper solution to this other than adding a 'view' tag to an existing ABI function which isn't a viable solution for those who load an existing abi or generate abi data from the .sol files?

### Steps to reproduce

Run above code in Nodejs + Typescript + Ethers V6 + Hedera JS SDK 2.35.0

### Hedera network

testnet

### Version

hashio testnet live

### Operating system

None

**Comment by AlfredoG87** [comment]
Hello @iron4548, I suspect that the error you are getting after the modification to your contract, `400 bad request` is due to ethers `v6` is trying to use a batch request (multiple RPC requests) at the same time.

This is a known limitation of the Relay and we are currently working on both supporting the **Batch Requests** Feature.

> To solve the issue, you can configure the JsonRpcProvider to serve a single eth method per request:
> 
>     const provider = new ethers.JsonRpcProvider(
>       `https://${networkId}.hashio.io/api`,
>       "",
>       {
>         batchMaxCount: 1,
>       }
>     );
> Please note using ethers v5 there is a clear distinction between an explicit batch-enabled rpcProvider ([JsonRpcBatchProvider](https://docs.ethers.org/v5/api/providers/other/#JsonRpcBatchProvider)) and a normal one ([UrlJsonRpcProvider](https://docs.ethers.org/v5/api/providers/other/#UrlJsonRpcProvider)), so you don't face that issue on ethers.js < 6.

**Reference Issues:** 
https://github.com/hashgraph/hedera-json-rpc-relay/issues/1802
https://github.com/hashgraph/hedera-json-rpc-relay/issues/1820

Let me know if this fixes your issue. and thank you for reporting an issue.

**Comment by iron4548** [comment]
Hi @AlfredoG87, thank you for your reply.

I am using your mentioned workaround, see above

```typescript
const networkId = 'testnet';
const provider = new ethers.JsonRpcProvider(`https://${networkId}.hashio.io/api`, '', {
  batchMaxCount: 1, //workaround for ethers V6
});
```

This code works fine with other function that has the 'view' tag, such as getAmountsIn and getAmountsOut from Uniswap V2. 

Both `quoteExactInput` and `quoteExactOutput` are intended to be used as a read-only operation, but they do not have a 'view' tag to them. 

I can confirm that those two functions works fine using code IF I manually add a 'view' tag to the ABI, i.e.

```
const interfaces = new ethers.Interface([
  'function quoteExactInput(bytes memory path, uint256 amountIn) external view returns (uint256 amountOut, uint160[] memory sqrtPriceX96AfterList, uint32[] memory initializedTicksCrossedList, uint256 gasEstimate)'
]);
```

But this isn't a suitable workaround (more of a hack). I'm writing sample codes for other devs to use, and they will be using their own generated ABI from the open source Uniswap V3 code.

Furthermore, if I use REST-API's contract call option then it works fine.

```typescript
 const interfaces = new ethers.Interface([
    'function quoteExactInput(bytes memory path, uint256 amountIn) external returns (uint256 amountOut, uint160[] memory sqrtPriceX96AfterList, uint32[] memory initializedTicksCrossedList, uint256 gasEstimate)'
  ]);

  const input =  100;
  const inputAmountInSmallestUnit =  BigNumber(input * Math.pow(10, config.aaaTokenDP));
  const token1 = TokenId.fromString(config.token1);
  const token2 = TokenId.fromString(config.token2);

  let pathHexStr = '0x';
  pathHexStr += tokenA.toSolidityAddress();  
  pathHexStr += feePercentToHexStr(0.03); //0.3%
  pathHexStr += tokenB.toSolidityAddress();

  const encodedPathData = hexToUint8Array(pathHexStr);
  const encodedData = interfaces.encodeFunctionData(interfaces.getFunction('quoteExactInput')!, [encodedPathData, inputAmountInSmallestUnit.toString()]);    

  const contractId = ContractId.fromString(config.quoter);

  const url = `${config.mirrorNodeBaseUrl}/api/v1/contracts/call`;
  const data = {
    'block': 'latest',
    'data': encodedData,
    'to': contractId.toSolidityAddress(),
  };

  const headers = {
    'content-type': 'application/json',
  };
 
 const response = await axios.post(url, data, { headers: headers });
 const resultHex = response.data.result; //works.
```

Bottom line, the issue is related to calling a read-only function via mirror node EVM that doesn't have the 'view' tag. The ethers V6 won't let me send the request because it expects a signer, so I added one. But then I got the 400 error.

**Comment by AlfredoG87** [comment]
sorry I oversaw the initialization and jump to soon to conclusions, since is the same `400 Bad Reques` 😓 

I am trying to replicate the issue, using the original snippet that you shared, but I got the issue that I don't have the configs that there are looking for:
`config.quoter`, `config.token1` and `config.token2`

Could you share those values to me please so I can complete the replication?

**Comment by iron4548** [comment]
> sorry I oversaw the initialization and jump too soon to conclusions, since is the same `400 Bad Request` 😓
> 
> I am trying to replicate the issue, using the original snippet that you shared, but I got the issue that I don't have the configs that there are looking for: `config.quoter`, `config.token1` and `config.token2`
> 
> Could you share those values to me please so I can complete the replication?

No worries. 

Let me get back to you regarding the configurations.

**Comment by iron4548** [comment]
@AlfredoG87, here's the configurations:

QuoterV2 contract id: 0.0.3172630
token 1 id: 0.0.143056
token 2 id:  0.0.143063

**Comment by AlfredoG87** [comment]
is it possible to share the project? or working script?

I just found 2 more missing parts of the code:

1. **feePercentToHexStr**
2. **hexToUint8Array**

**Comment by iron4548** [comment]
Here are the missing functions.

````typescript

function feePercentToHexStr(feePercent: number): string {
  const feeBips = Math.round(feePercent * 100000);
  const buffer = new Uint8Array([
    (feeBips >> 16) & 0xFF,
    (feeBips >> 8) & 0xFF,
    feeBips & 0xFF
  ]);

  return Array.from(buffer)
    .map(byte => byte.toString(16).padStart(2, '0'))
    .join('');
}

function hexToUint8Array(hexString: string): Uint8Array {
  hexString = hexString.startsWith('0x') ? hexString.slice(2) : hexString;
  return new Uint8Array(hexString.match(/.{1,2}/g)!.map(byte => parseInt(byte, 16)));
}
```

**Comment by iron4548** [comment]
Hi @AlfredoG87, I was able to find a workaround. I was able to bypass the 'view' check internally by executing the 'call' method via the provider object instead as shown below.

I think no further action is needed here and this issue can be closed.

```typescript

  const data = abiInterfaces.encodeFunctionData('quoteExactInput', [encodedPathData, inputAmountInSmallestUnit.toString()]);

  //send a call to the JSON-RPC provider directly
  const result = await provider.call({
    to: quoterEvmAddress,
    data: data,
  });

  const decoded = abiInterfaces.decodeFunctionResult('quoteExactInput', result); 
  const finalOutputAmount = decoded.amountOut; //in token's smallest unit

```

**Comment by AlfredoG87** [comment]
Hi @iron4548, am so glad you were able to find a solution 💯
